src/functional/sim_piece.zig

```diff
@@ -0,0 +1,765 @@
+        if (!hash_to_hash_result.found_existing) {
+            hash_to_hash_result.value_ptr.* = HashMap(
+                f64,
+                ArrayList(usize),
+                HashF64Context,
+                std.hash_map.default_max_load_percentage,
+            ).init(allocator);
+        }
+        // Get or put the inner HashMap entry for the given slope
+        const hash_to_array_result = try hash_to_hash_result.value_ptr.*.getOrPut(slope);
+        if (!hash_to_array_result.found_existing) {
+            hash_to_array_result.value_ptr.* = ArrayList(usize).init(allocator);
+        }
+        try hash_to_array_result.value_ptr.*.append(segment_metadata.start_time);
+    }
+}
+
+/// Sim-Piece Phase 4. Create from the `merged_segments_metadata_map` and byte array that can be
+/// decoded during decompression and store it in `compressed_values`.
+fn createCompressedRepresentation(
+    merged_segments_metadata_map: HashMap(f64, HashMap(
+        f64,
+        ArrayList(usize),
+        HashF64Context,
+        std.hash_map.default_max_load_percentage,
+    ), HashF64Context, std.hash_map.default_max_load_percentage),
+    compressed_values: *ArrayList(u8),
+) !void {
+    // Iterate over the hash to hash map.
+    var hash_to_hash_iterator = merged_segments_metadata_map.iterator();
+    while (hash_to_hash_iterator.next()) |hash_to_hash_entry| {
+        const current_interception: f64 = hash_to_hash_entry.key_ptr.*;
+        // Append the `current_interception`.
+        try appendValue(f64, current_interception, compressed_values);
+
+        // Append the number of slopes hashed by the `current_interception`.
+        try appendValue(usize, hash_to_hash_entry.value_ptr.*.count(), compressed_values);
+        var hash_to_array_iterator = hash_to_hash_entry.value_ptr.*.iterator();
+
+        while (hash_to_array_iterator.next()) |hash_to_array_entry| {
+            const current_slope = hash_to_array_entry.key_ptr.*;
+            // Append the `current_slope` that approximates the segment.
+            try appendValue(f64, current_slope, compressed_values);
+
+            // Append the number of timestamps that the `current_slope` is hashing.
+            try appendValue(usize, hash_to_array_entry.value_ptr.*.items.len, compressed_values);
+            var previous_timestamp: usize = 0;
+            // Iterate over the timestamps and append them.
+            for (hash_to_array_entry.value_ptr.*.items) |timestamp| {
+                try appendValue(usize, timestamp - previous_timestamp, compressed_values);
+                previous_timestamp = timestamp;
+            }
+        }
+    }
+}
+
+/// Quantizes the given `value` by the specified `error_bound`. This process ensures that
+/// the quantized value remains within the error bound of the original value. If the
+/// `error_bound` is equal to zero, the value is directly returned.
+fn quantize(value: f64, error_bound: f32) f64 {
+    if (error_bound != 0)
```

Comment:
Maybe add curly braces around the true block of the if expression for consistency and to minimize the chance of introducing bugs if the code is changed.

Reply:
Of course it depends on how annoyingly pedantic the reviewer is :grin:

In my opinion there is very little benefit to not write the curly braces and if they are always included the code becomes more consistent and it becomes more clear what lines of code are included as part of the branch. Others seems to have come to the same conclusion as the C++ style guides I am familiar with also requires that curly braces are always included except for very specific situations, e.g., see [Google's C++ Style Guide](https://google.github.io/styleguide/cppguide.html#Formatting_Looping_Branching).